kvui.py

```diff
@@ -299,6 +300,39 @@ def apply_selection(self, rv, index, is_selected):
         self.selected = is_selected
 
 
+class AutocompleteHintInput(TextInput):
+    min_chars = NumericProperty(3)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self.dropdown = DropDown()
+        self.dropdown.bind(on_select=lambda instance, x: setattr(self, 'text', x))
+        self.bind(on_text_validate=self.on_message)
+
+    def on_message(self, instance):
+        App.get_running_app().commandprocessor("!hint "+instance.text)
+
+    def on_text(self, instance, value):
+        if len(value) >= self.min_chars:
+            self.dropdown.clear_widgets()
+            ctx: context_type = App.get_running_app().ctx
+            item_names = ctx.item_names._game_store[ctx.game].values()
+
+            def on_press(button: Button):
+                return self.dropdown.select(button.text)
+            lowered = value.lower()
+            for item_name in item_names:
+                if lowered in item_name.lower():
+                    btn = Button(text=item_name, size_hint_y=None, height=dp(30))
```

Comment:
```suggestion
                    highlighted_item_name = re.sub(
                        value,
                        lambda m: f'[color={JSONtoTextParser.color_codes["orange"]}]{m.group()}[/color]', item_name, flags=re.IGNORECASE
                    )
                    btn = Button(text=highlighted_item_name, size_hint_y=None, height=dp(30), markup=True)
```
You can get a cool effect highlighting the parts that the search found pretty easily:
![image](https://github.com/ArchipelagoMW/Archipelago/assets/31861583/d3ba3e53-23a7-45eb-b563-63496bb5cf5c)

Not too attached to this idea, so cool if you drop it as well. I could see it making fuzzy search harder.

Reply:
Note: this breaks selection as the markup tags will appear in the text box when it's selected. I've solved that in #3537 with an extra field on the button class to store the un-marked-up item name.

Reply:
should be solvable with kivy's own markup escape.